File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/PurchaseEntitlementUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.repository.EntitlementCacheWriter
import com.izzatismail.midtrim.domain.repository.PlayBillingService
import com.izzatismail.midtrim.domain.repository.PurchaseResult
import java.util.Date

class PurchaseEntitlementUseCase(
    private val billingService: PlayBillingService,
    private val cache: EntitlementCacheWriter
) {
    private val productId = "com.midtrim.fullunlock"

    suspend fun execute(): PurchaseResult {
        val result = billingService.purchase(productId)
        if (result is PurchaseResult.Success) {
            cache.isPurchased = true
            cache.productId = productId
            cache.lastVerifiedAt = Date().time
        }
        return result
    }
}

File: ios/MidTrim/MidTrim/Domain/UseCases/PurchaseEntitlementUseCase.swift
import Foundation

struct PurchaseEntitlementUseCase {
    private let storeService: StoreKitServiceProtocol
    private let cache: EntitlementCacheProtocol
    private let productID: String

    init(storeService: StoreKitServiceProtocol, cache: EntitlementCacheProtocol, productID: String = "com.midtrim.fullunlock") {
        self.storeService = storeService
        self.cache = cache
        self.productID = productID
    }

    func execute() async -> PurchaseResult {
        let result = await storeService.purchase(productID: productID)
        if case .success = result {
            await cache.setPurchased(true)
            await cache.setProductId(productID)
            await cache.setLastVerified(Date())
        }
        return result
    }
}
